src/main/java/com/marklogic/client/ext/datamovement/consumer/WriteRowToTableauConsumer.java

```diff
@@ -0,0 +1,250 @@
+            if ( columns.size() == 0 ) {
+              for (String columnName : vals.keySet()) {
+                XsAnyAtomicTypeVal value = vals.get(columnName);
+                if (value instanceof XsBooleanVal) {
+                  withColumn(columnName, Type.BOOLEAN);
+                } else if (value instanceof XsStringVal) {
+                  withColumn(columnName, Type.UNICODE_STRING);
+                } else if (value instanceof XsIntegerVal ||
+                  value instanceof XsIntVal ||
+                  value instanceof XsShortVal ||
+                  value instanceof XsLongVal ) {
+                  withColumn(columnName, Type.INTEGER);
+                } else if (value instanceof XsFloatVal ||
+                  value instanceof XsDoubleVal ) {
+                  withColumn(columnName, Type.DOUBLE);
+                } else {
+                  throw new IllegalStateException("Unsupported type: " + value.getClass().getName());
+                }
+              }
+            }
+            for ( Column column : columns.values() ) {
+              if ( column.collation != null ) {
+                tableDef.addColumnWithCollation(column.columnName,
+                  column.type, column.collation);
+              } else {
+                tableDef.addColumn(column.columnName, column.type);
+              }
+            }
+            // as of version 10.3.7 Tableau Data Extact SDK only supports
+            // one table named "Extract"
+            table = extract.addTable("Extract", tableDef);
+            initialized = true;
+          } catch (TableauException e) {
+            throw new MarkLogicIOException(e);
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Add a column to the Consumer to add it in the Tableau Data Extract file
+   *
+   * @param columnName the column name specified in the Template Driven
+   *          Extraction template (the column names are case sensitive)
+   * @param type the Tableau Data Type
+   * @return the Consumer for chaining
+   */
+  public WriteRowToTableauConsumer withColumn(String columnName, Type type) {
+    checkSupportedTypes(type);
+    columns.put(columnName, new Column(columnName, type));
+    return this;
+  }
+
+  /**
+   * Add a column with tableau collation to the Consumer to add it in the
+   * Tableau Data Extract file
+   *
+   * @param columnName the column name specified in the Template Driven
+   *          Extraction template (the column names are case sensitive)
+   * @param type the Tableau Data Type
+   * @param collation the Tableau Data Collation
+   * @return the Consumer for chaining
+   */
+  public WriteRowToTableauConsumer withColumn(String columnName, Type type, Collation collation) {
+    checkSupportedTypes(type);
+    columns.put(columnName, new Column(columnName, type, collation));
+    return this;
+  }
+
+  // closes underlying TableDefinition and Extract. This should be called by the
+  // QueryBatchListener
+  public void close() {
```

Comment:
Is this part of some interface that ExtractViaTemplateListener depends on? Presumably AutoCloseable? If so, it's good to have an @Override annotation on it.